### PR TITLE
CLN: amended TypeError from pd.concat() to be more informative (GH15796)

### LINKED ISSUE #15796: ERR: better error message for trying to concat non-pandas objects
```
In [3]: pd.concat([Series(), 'foo'])
TypeError: cannot concatenate a non-NDFrame object
```

trivially ``non-NDFrame`` -> ``non-pandas``. but I think even better would be something like

```
In [4]: objs = [Series(), 'foo']

In [7]: "cannot concatenate [{pandas}] with non-pandas objects of type [{non_pandas}]".format(
   ...: pandas=','.join(set([str(o.__class__.__name__) for o in objs if isinstance(o, pandas.core.generic.NDFrame)])),
   ...: non_pandas=','.join(set([str(type(o)) for o in objs if not isinstance(o, pandas.core.generic.NDFrame)])))
Out[7]: "cannot concatenate [Series] with non-pandas objects of type [<class 'str'>]"
```

### PR DESCRIPTION
**Pull request by AdamShamlian**
CLN/TST: Added regex matching for exception message

- [X] closes #15796 
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

**Comment by AdamShamlian** [comment]
Hey Pandas team. Long time user here looking to start giving back. I've been burned by this error message plenty in my early days, so a fitting first issue.

File: pandas/core/reshape/concat.py [hunk 1]
```diff
@@ -262,7 +262,13 @@ def __init__(self, objs, axis=0, join='outer', join_axes=None,
         ndims = set()
         for obj in objs:
             if not isinstance(obj, NDFrame):
-                raise TypeError("cannot concatenate a non-NDFrame object")
+
+                # multiline string concat is safer with whitespace for regex
+                # at test_concat.py:181
+                msg = """cannot concatenate object of type "{0}";"""
+                msg += " only pd.Series, pd.DataFrame, and pd.Panel"
+                msg += " objs are valid".format(type(obj))
```
Comment (jorisvandenbossche) [inline pandas/core/reshape/concat.py]: Can you write this as 

```
msg = (" ..."
       " ... ")
```

(which is the typical way we write multi-line strings in the codebase)

File: pandas/tests/reshape/test_concat.py [hunk 2]
```diff
@@ -177,7 +177,15 @@ def test_concatlike_same_dtypes(self):
...
-            msg = "cannot concatenate a non-NDFrame object"
+
+            # multiline string concat is safer with \n and whitespace
+            # for regex (see concat.py:266
+
+            # (.+?) may be too general but we are more concerned
+            # with the message as a whole, anyways
+            msg = """cannot concatenate object of type \"(.+?)\";"""
+            msg += " only pd.Series, pd.DataFrame, and pd.Panel"
+            msg += " objs are valid"
```
Comment (jorisvandenbossche) [inline pandas/tests/reshape/test_concat.py]: same comment here as above. 
But also, I think it is fine to only match a part of the error message if that makes it easier.

File: pandas/core/reshape/concat.py [hunk 3]
```diff
@@ -262,7 +262,13 @@ def __init__(self, objs, axis=0, join='outer', join_axes=None,
         ndims = set()
         for obj in objs:
             if not isinstance(obj, NDFrame):
-                raise TypeError("cannot concatenate a non-NDFrame object")
+
+                # multiline string concat is safer with whitespace for regex
+                # at test_concat.py:181
```
Comment (jorisvandenbossche) [inline pandas/core/reshape/concat.py]: I am not fully understanding this comment

**Review by jorisvandenbossche — COMMENTED** [review]
Thanks for starting to contribute! 
I added a few comments

File: pandas/core/reshape/concat.py [hunk 4]
```diff
@@ -262,7 +262,13 @@ def __init__(self, objs, axis=0, join='outer', join_axes=None,
...
+                msg = """cannot concatenate object of type "{0}";"""
+                msg += " only pd.Series, pd.DataFrame, and pd.Panel"
```
Comment (jreback) [inline pandas/core/reshape/concat.py]: no need to mention ``pd.Panel`` as they are deprecated

File: pandas/tests/reshape/test_concat.py (hunk 2 above)
Comment (AdamShamlian) [reply pandas/tests/reshape/test_concat.py]: Thanks, I'll clean this up (which will eliminate the confusing comment as well).

File: pandas/core/reshape/concat.py (hunk 4 above)
Comment (AdamShamlian) [reply pandas/core/reshape/concat.py]: Ahh, were there plans to maintain it in some "LTS" pre 1.0 release, or is it full-on deprecated?

Comment (jreback) [reply pandas/core/reshape/concat.py]: are you offering to do this ???
haven’t really decided what to do with this yet

Comment (AdamShamlian) [reply pandas/core/reshape/concat.py]: No, that sounds decidedly outside the scope of the time I have right now. While I was aware of the deprecation, I thought there was discussion about it being passively maintained. I don't have any horses in the pd.Panel race. Just curious.

File: pandas/core/reshape/concat.py [hunk 5]
```diff
@@ -262,7 +262,10 @@ def __init__(self, objs, axis=0, join='outer', join_axes=None,
         ndims = set()
         for obj in objs:
             if not isinstance(obj, NDFrame):
-                raise TypeError("cannot concatenate a non-NDFrame object")
+                msg = ('cannot concatenate object of type "{0}";'
```
Comment (jreback) [inline pandas/core/reshape/concat.py]: actually lets add back Panel, they are current still valid (same below)

Comment (AdamShamlian) [reply pandas/core/reshape/concat.py]: Haha, ok. Should I mark it as "... pd.Panel (deprecated)" or leave out the deprecated note?

Comment (jreback) [reply pandas/core/reshape/concat.py]: I like that first one! (mark as deprecated)

Comment (AdamShamlian) [reply pandas/core/reshape/concat.py]: Should be good to go now

Comment (AdamShamlian) [reply pandas/core/reshape/concat.py]: Let that last message be living testament to my hubris. Forgot to escape the parens for regex and busted the build system. NOW it should be good to go.

**Comment by jreback** [comment]
thanks @AdamShamlian

**Pull request merged**